chiselFrontend/src/main/scala/chisel3/RawModule.scala

```diff
@@ -134,6 +134,14 @@ abstract class RawModule(implicit moduleCompileOptions: CompileOptions)
   }
 }
 
+trait RequireAsyncReset extends MultiIOModule {
+  override private[chisel3] def mkReset: AsyncReset = AsyncReset()
+}
+
+trait RequireSyncReset extends MultiIOModule {
+  override private[chisel3] def mkReset: Bool = Bool()
+}
+
```

Comment:
Is there no possibility for something like:

```scala
trait RequireSpecificResetType extends MultiIOModule {
  def resetTypeGen: () => Reset
  override private[chisel3] def mkReset: Reset = resetTypeGen()
}
```

This would later let someone do something like:

```scala
val MySpecifiedResetTypeModule(useAsyncReset: Bool) extends Module with RequireSpecificResetType {
  override def resetTypeGen() = if (useAsyncReset) AsyncReset() else Bool()
  ...
}
```

Reply:
Would a companion object be an alternative that would work for you, @mwachs5?

```scala
object Foo { 
  def apply(useAsyncReset: Boolean) = useAsyncReset match {
    case true  => new Foo with RequireAsyncReset
    case false => new Foo with RequireSyncReset
  }
}
```

Barring that being insufficient, I think it's reasonable (as @jackkoenig suggests) to open up `mkReset` as `protected`.

Reply:
Hm, that could work, will try it out

Reply:
So trying this out it's still a little clunky. Specifically rocket-chip generation flow needs a top level Module and a Config class. So if i want my top level module fixed but the Config class to have information about the reset type for the top level, this is hard to express with a companion object or a fixed reset scheme. I guess i can change the top level module itself but this is rather unnatural.

Reply:
Yes I really think this is rather limiting. Can I request to have the `mkReset` be protected?